Bank: Add function to replace empty account with upgradeable program on feature activation

buffalojoec opened the pull request:
#### Problem
Currently, `replace_program_account` will only replace a program's account, but not the program's **data account**.

In order to complete Step 1 of #32780, we need to be able to replace an empty account with an upgradeable program on feature activation.

#### Summary of Changes
This PR adds a function `replace_empty_account_with_upgradeable_program` to replace an empty account - such as `Feature111111111` with a program account and a data account, effectively replacing this account with an upgradeable program.

buffalojoec committed:
replace program account

mvines commented:
hey @buffalojoec, you should get triage and than write access to the monorepo per https://github.com/solana-labs/contributor-access-policy so that CI will run on your PRs automatically. I'll vouch for ya and I'm pretty sure Jon will too

joncinque reviewed (commented):
Now that I'm thinking through what would happen, I'm worried that this would break things unfortunately.

Let's imagine we have an upgradeable program on address `A`, and we want to replace it with an upgradeable program on `B`. Here's the initial situation:

* `A`'s data is `PDA(A)`
* `PDA(A)`'s data is the old program
* `B`'s data is `PDA(B)`
* `PDA(B)`'s data is the new program

If we replace `A` with `B`, and `PDA(A)` with `PDA(B)`, then we get:

* `A`'s data is just `PDA(B)` <-- Uh oh!
* `PDA(A)`'s data is the new program
* `B` is gone
* `PDA(B)` is gone

So if we want to replace an upgradeable program with another, we *only* need to replace the account at `PDA(A)` with `PDA(B)`.

The feature-gate situation is a little bit strange though, since we want to *create* a whole new program at `Feature111111`. I'm thinking you *will* need to replace `A` with `B`, but you'll *also* need to rewrite the data at `A` to be `PDA(A)`.

Let me know if this makes sense. Could you add some more to the test to reflect the data written into the accounts to check the situation that I'm worried about?

buffalojoec commented:
> Let me know if this makes sense. Could you add some more to the test to reflect the data written into the accounts to check the situation that I'm worried about?

Yeah, I think that makes sense. In other words, replace the data account instead of the program account. 

I guess what's more confusing is how this even works to upgrade programs in the current state, without the data account being involved 🧐

Regardless, happy to revise and write some more testing !

joncinque commented:
> Yeah, I think that makes sense. In other words, replace the data account instead of the program account.

Yes, *and also* rewrite the program account with the correct address if needed.

> I guess what's more confusing is how this even works to upgrade programs in the current state, without the data account being involved 🧐

It only works with non-upgradeable programs, which just have the one account :wink:

buffalojoec commented:
I just pushed a modified method and some fleshed out tests evaluating your points above @joncinque. Let me know if this aligns with what you were thinking.

TL/DR: The `test_non_upgradable_program_replace_with_data` test is the use case for the Feature Gate program to be deployed at `Feature111111`.

I've added verbose comments in order to fuel discussion about how to handle each of the four cases identified in the changes.

buffalojoec committed:
modify for all cases

buffalojoec commented:
I just pushed a modified method and some fleshed out tests evaluating your points above @joncinque. Let me know if this aligns with what you were thinking.

TL/DR: The `test_non_upgradable_program_replace_with_data` test is the use case for the Feature Gate program to be deployed at `Feature111111`.

I've added verbose comments in order to fuel discussion about how to handle each of the four cases identified in the changes.

buffalojoec commented:
I wanted to create this updated comment to flesh out some of the logic articulated by [d646a07](https://github.com/solana-labs/solana/pull/32783/commits/d646a076547b123f43c347a8152939f8cdea641e).

There are four scenarios where we want to consider how to handle program replacements, now that we're involving the data account as well. Each of these four scenarios has a corresponding test and is handled by some conditional block in the `replace_program_account` function itself.

### Definitions

- `existence`: Account currently exists on-chain and is not empty
- `old`: The `existence` of these accounts defines the program's current state
- `new`: The `existence` of these accounts serves to replace the associated current state at these accounts' counterparts

### Scenarios

#### Scenario 1: Program Account Swap

The existing program and the proposed new program _both_ just have program accounts, no data accounts.

Swap the existing program account for the new one.

| Current      | Proposed | Result |
| ----------- | ----------- | ---------- |
| `Old:     [Old program data]` | `New:     [*New program data]` |  `Old:     [*New program data]` |

#### Scenario 2: Program Account Swap and Create Data

The existing program does not have a data account, but the proposed new one does.

Swap the existing program account for the PDA of the new data account, and create the new data account.

| Current      | Proposed | Result |
| ----------- | ----------- | ---------- |
| `Old:     [Old program data]` | `New:     PDA(NewData)` |  `Old:     PDA(OldData)` |
|  | `NewData: [*New program data]` |  `OldData: [*New program data]` |

#### Scenario 3: Program Account Swap and Remove Data

The existing program has a data account, but the proposed new one _does not_.

Swap the existing program account for the new program account, and remove the existing data account.

| Current      | Proposed | Result |
| ----------- | ----------- | ---------- |
| `Old:     PDA(OldData)` | `New:     [*New program data]` |  `Old:     [*New program data]` |
| `OldData: [Old program data]` | | |

#### Scenario 4: Program Account Swap

The existing program and the proposed new program _both_ have data accounts.

Swap the existing data account for the new one, leaving the existing program account (with the PDA of the data account) unchanged.

| Current      | Proposed | Result |
| ----------- | ----------- | ---------- |
| `Old:     PDA(OldData)` | `New:     PDA(NewData)` |  `Old:     PDA(OldData)` |
| `Old:     [Old program data]` | `New:     [*New program data]` |  `Old:     [*New program data]` |

buffalojoec commented:
After offline discussion with @joncinque I've removed Scenario 3 from the supported operations!

buffalojoec committed:
remove non-data swap

joncinque commented on runtime/src/bank.rs:
```diff
@@ -8115,30 +8135,81 @@ impl Bank {
+                            None,
+                            &new_data_account,
+                        );
+                        // Update the old program account to house the PDA of
+                        // the data account
+                        {
+                            let mut account = Account::from(old_account.clone());
+                            account.data = old_data_address.as_ref().to_vec();
+                            let account_shared_data = AccountSharedData::from(account);
+                            self.store_account(old_address, &account_shared_data);
+                        }
```
This account might not have any data, meaning probably no lamports for rent exemption, so we need to make sure that it has the lamports or the runtime will garbage collect it.

I think we'll need to take the lamports from the new program account.

joncinque commented on runtime/src/bank.rs:
```diff
@@ -8115,30 +8135,81 @@ impl Bank {
+                        {
+                            let mut account = Account::from(old_account.clone());
+                            account.data = old_data_address.as_ref().to_vec();
+                            let account_shared_data = AccountSharedData::from(account);
```
nit: is this line needed? it seems like `Account` should be storable no problem, but I might be missing something

joncinque commented on runtime/src/bank.rs:
```diff
@@ -8115,30 +8135,81 @@ impl Bank {
+                            self.store_account(old_address, &account_shared_data);
+                        }
+                    }
+                    // We only swapped the data accounts, so now we need to
+                    // clear the new program account
+                    self.capitalization
+                        .fetch_sub(new_account.lamports(), Relaxed);
```
Heh, *maybe*

buffalojoec replied:
Only the proposed new data account was used, so do we leave this account alone or clear it? After all, it should just be a pointer to the now-deceased data account, which is now in place of the old program data account

joncinque replied:
Ah sorry, this was in reference to my previous comment -- if we end up stealing the lamports from the previous account, we don't want to subtract those from the total capitalization.

joncinque commented on runtime/src/bank/tests.rs:
```diff
@@ -8013,44 +8013,208 @@ fn test_compute_active_feature_set() {
-    bank.replace_program_account(&old_address, &new_address, "bank-apply_program_replacement");
+    bank.replace_program_account(&old, &new, "bank-apply_program_replacement");
+
+    // Old program account balance is unchanged
+    assert_eq!(bank.get_balance(&old), 100);
```
Shouldn't it have the balance of the new account? If we replace an old program with a new bigger program, the old address should now have the balance from the new address, no?

joncinque commented on runtime/src/bank/tests.rs:
```diff
@@ -8013,44 +8013,208 @@ fn test_compute_active_feature_set() {
+    let old_account = bank.get_account(&old).unwrap();
+    assert_eq!(old_account.data(), &new_program_bytes,);
+
+    // Lamports in the old token account was burnt
+    assert_eq!(bank.capitalization(), original_capitalization - 100);
```
This seems to hint that the behavior might be correct?

joncinque commented on runtime/src/bank.rs:
```diff
@@ -8115,30 +8135,81 @@ impl Bank {
+                    );
+                }
 
                 // Unload a program from the bank's cache
                 self.loaded_programs_cache
                     .write()
                     .unwrap()
                     .remove_programs([*old_address].into_iter());
```
We should also test to make sure this doesn't have any issues if the old program didn't exist before. I'm imagining we just need to skip it and we'll be good

joncinque commented on runtime/src/bank/tests.rs:
```diff
@@ -8013,44 +8013,208 @@ fn test_compute_active_feature_set() {
+    lamports: u64,
+    data: Vec<u8>,
+) -> AccountSharedData {
+    let new_account = AccountSharedData::from(Account {
+        lamports,
+        data,
+        ..Account::default()
+    });
```
It might be good to also take the owner and executable flag into account for a full test, since those will change for the feature gate program replacement.

joncinque commented on runtime/src/bank/tests.rs:
```diff
@@ -8013,44 +8013,208 @@ fn test_compute_active_feature_set() {
 
-    // Lamports in the old token account were burnt
+    // Old program account balances are unchanged
+    assert_eq!(bank.get_balance(&old), 100);
```
This should be changed, but the test isn't detecting it because `old` and `new` both have `100`. Can you change the number to make sure that `old` gets the balance from `new`?

joncinque commented on runtime/src/bank/tests.rs:
```diff
@@ -8013,44 +8013,208 @@ fn test_compute_active_feature_set() {
+    let old_data_account = bank.get_account(&old_data).unwrap();
+    assert_eq!(old_data_account.data(), &new_program_bytes,);
+
+    // Lamports in the old token accounts were burnt
```
micro-nit: lol oops, probably my mistake
```suggestion
    // Lamports in the old program accounts were burnt
```

joncinque commented on runtime/src/bank/tests.rs:
```diff
@@ -8013,44 +8013,208 @@ fn test_compute_active_feature_set() {
+
+    let old = Pubkey::new_unique();
+    let (old_data, _) = Pubkey::find_program_address(&[old.as_ref()], &bpf_id);
+    set_up_account_with_bank(&mut bank, &old, 100, old_data.to_bytes().to_vec());
+    set_up_account_with_bank(&mut bank, &old_data, 102, vec![0, 1, 2, 3, 4, 5, 6]);
+
+    let new = Pubkey::new_unique();
+    let (new_data, _) = Pubkey::find_program_address(&[new.as_ref()], &bpf_id);
+    let new_program_bytes = vec![6, 5, 4, 3, 2, 1, 0];
+    set_up_account_with_bank(&mut bank, &new, 100, new_data.to_bytes().to_vec());
+    set_up_account_with_bank(&mut bank, &new_data, 102, new_program_bytes.clone());
```
Let's use different lamport numbers here to be sure the test is correct

joncinque commented on runtime/src/bank/tests.rs:
```diff
@@ -8013,44 +8013,208 @@ fn test_compute_active_feature_set() {
+    let old = Pubkey::new_unique();
+    let (old_data, _) = Pubkey::find_program_address(&[old.as_ref()], &bpf_id);
+    let old_program_bytes = vec![0, 1, 2, 3, 4, 5, 6];
+    set_up_account_with_bank(&mut bank, &old, 100, old_data.to_bytes().to_vec());
```
And finally, let's make these lamport values unique too to be sure

joncinque reviewed (commented):
This is on the right track, thanks for all the comments and tests. Now we just need to make sure it'll cover the feature gate program situation

buffalojoec commented on runtime/src/bank.rs:
```diff
@@ -8115,30 +8135,81 @@ impl Bank {
+                        self.get_account_with_fixed_root(&old_data_address)
+                    {
+                        // It does. Replace it with the new data account
+                        self.replace_account(
+                            &old_data_address,
+                            &new_data_address,
+                            Some(&old_data_account),
+                            &new_data_account,
+                        );
```
Sorry, I left this commit purposely verbose in some areas to make sure the logic was correct!

buffalojoec commented on runtime/src/bank.rs:
```diff
@@ -8115,30 +8135,81 @@ impl Bank {
+                            Some(&old_data_account),
+                            &new_data_account,
+                        );
+                        // The old program account will already house the PDA
+                        // of the data account
```
I wasn't sure if this was the desired logic. For example, if you have just a program with no data account, and it has some arbitrary data in it, do you want to be able to wipe it for a new program and data account? ie. effectively changing it from non-upgradeable to upgradeable?

buffalojoec replied:
Sure makes sense! I kind of assumed it would be if there was a data account already

buffalojoec commented on runtime/src/bank.rs:
```diff
@@ -8115,30 +8135,81 @@ impl Bank {
         new_address: &Pubkey,
         datapoint_name: &'static str,
     ) {
+        // Both program accounts must exist in order to attempt a replacement
```
OK I can do that. I also assumed we could send lamports to it ahead of time, but I'll work that in

buffalojoec committed:
address tests & conditional feedback

buffalojoec committed:
get the rent involved

buffalojoec committed:
mix in owner & executable

buffalojoec committed:
feature-related cases

buffalojoec commented:
Just pushed up some commits addressing your feedback.

A few questions came to mind while working on them:
- Do we want to enforce the proper BPF loader owns a "new" account? ie. if we are tasked with replacing a program and data account with another program and data account, should we have any checks that those two new accounts are owned by the upgradable loader?
- Do we want to enforce that a new program account contains the PDA for it's associated data account? Despite the fact that when both are proposed as replacements, the new program account is all but completely ignored
- Do we want to keep excess lamports in accounts or burn them? If either type of account is replaced with smaller data, what should happen to those lamports that are excess of rent-exemption?

joncinque commented:
> Do we want to enforce the proper BPF loader owns a "new" account? ie. if we are tasked with replacing a program and data account with another program and data account, should we have any checks that those two new accounts are owned by the upgradable loader?

We just need to be sure that we're setting the owner that's in the new account, we don't need to check it necessarily. The issue to watch out for is swapping accounts that are on different loaders, and *not* setting the old account to the new program's loader.

> Do we want to enforce that a new program account contains the PDA for it's associated data account? Despite the fact that when both are proposed as replacements, the new program account is all but completely ignored

We could simply read the data from the new program account and use that rather than every dealing with the new PDA, but it's up to you. Ignoring is probably fine.

> Do we want to keep excess lamports in accounts or burn them? If either type of account is replaced with smaller data, what should happen to those lamports that are excess of rent-exemption?

Burn 'em! no need to have more lamports than necessary

Side note: It's funny, I'm just realizing that scenario 3 is possible with loader v4, but we don't need to address it here.

joncinque commented:
Looking through the code and all the cases, I'm starting to get cold feet about a generic replace function. This code is all really dangerous since it goes around normal consensus to move data and lamports around, so it should be as simple and limited as possible. All of the possible branches make me worry that we're going to miss a case and end up taking down the network because the capitalization is wrong.

What do you think about keeping the refactor that you've done with `replace_account` and creating a new function like `replace_empty_account_with_upgradeable_program` for the Feature Gate program? Maybe later we can add a more general `replace_program`, which can call into the more specific versions as needed.

I do appreciate all of the great work and all of the tests, but I think we should be as precise as possible to support only the cases that we need to. What do you think?

buffalojoec commented:
> What do you think about keeping the refactor that you've done with `replace_account` and creating a new function like `replace_empty_account_with_upgradeable_program` for the Feature Gate program? Maybe later we can add a more general `replace_program`, which can call into the more specific versions as needed.

Yeah it got sketchy _real_ quick with all the different scenarios.

Where is this function invoked anyway? I guess I didn't realize creating a new separate bank method was on the table

joncinque commented:
> Where is this function invoked anyway? I guess I didn't realize creating a new separate bank method was on the table

You can see how it's used in https://github.com/solana-labs/solana/pull/31764 in `apply_feature_activations`, you'll need to add a new feature gate which replaces `Feature11111...` with another program address that contains a no-op program

buffalojoec committed:
stripped back to feature-specific case only

buffalojoec committed:
added feature

buffalojoec commented:
This should be ready for another look. Let me know if I should squash and put up a new PR.

No-op program:
- `5Re2FUHmvSdem1KV2bU5GFvqZpP2MXj76yVhMKKbP45o`
- [Explorer](https://explorer.solana.com/address/5Re2FUHmvSdem1KV2bU5GFvqZpP2MXj76yVhMKKbP45o)
- [Repository](https://github.com/buffalojoec/noop-programs/tree/main/feature-gate-noop)

Feature:
- `8GdovDzVwWU5edz2G697bbB7GZjrUc6aQZLWyNNAtHdg`

(It also doesn't look like that failed CI was me)

joncinque reviewed (commented):
This was so much easier to review, thanks a ton for reducing the scope!

I looked at the program you uploaded, and even though the extend program instruction is meant to go out with in 1.16, could you add some huge static string to make the size bigger? That way we can't get stuck with a program data account that's too small.

Also, is it a verifiable build? I'd like it to be possible for anyone in the multisig to check it to make sure it's correct before signing off on an upgrade.

buffalojoec committed:
address initial feedback

buffalojoec committed:
added more lamport checks

buffalojoec commented on runtime/src/bank.rs:
```diff
@@ -8131,26 +8170,68 @@ impl Bank {
-                    old_account.data().len(),
-                    new_account.data().len(),
+                let data = old_data_address.as_ref().to_vec();
+                let lamports = self.get_minimum_balance_for_rent_exemption(data.len());
```
What exactly do you mean by "create new"?

`get_minimum_balance_for_rent_exemption` just calculates the rent due, and the change in capitalization is subtracting them from the new account's lamports, which is destroyed in the process, effectively burning the remaining lamports of the new account after rent is subtracted.

In other words, I set this up so no lamports would be created, but instead taken from the new account which is about to be wiped

joncinque replied:
Sorry, I meant to say "has the possibility to create new lamports if the new program account doesn't have any", but that should never happen because of what you do later

buffalojoec replied:
Ohh yep I added a commit for tighter lamport checks, pushing soon

buffalojoec committed:
condense tests

buffalojoec committed:
using test_case

buffalojoec commented:
> I looked at the program you uploaded, and even though the extend program instruction is meant to go out with in 1.16, could you add some huge static string to make the size bigger? That way we can't get stuck with a program data account that's too small.

Sure thing, how huge are you thinking? What's the ideal safe spot? We must choose wisely...

> Also, is it a verifiable build? I'd like it to be possible for anyone in the multisig to check it to make sure it's correct before signing off on an upgrade.

I can make it one!

joncinque commented:
> Sure thing, how huge are you thinking? What's the ideal safe spot? We must choose wisely...

Let's give ourselves 150kb? That's the size of the token program, which seems reasonable. What do you think?

> I can make it one!

Thank you!

buffalojoec committed:
add fail cases to tests

buffalojoec commented:
Resolved all comments you shared. I also added some other tests

> > I looked at the program you uploaded, and even though the extend program instruction is meant to go out with in 1.16, could you add some huge static string to make the size bigger? That way we can't get stuck with a program data account that's too small.
> 
> Sure thing, how huge are you thinking? What's the ideal safe spot? We must choose wisely...
> 
> > Also, is it a verifiable build? I'd like it to be possible for anyone in the multisig to check it to make sure it's correct before signing off on an upgrade.
> 
> I can make it one!

buffalojoec commented:
> Let's give ourselves 150kb? That's the size of the token program, which seems reasonable. What do you think?

I guess I can't foresee the feature gate program being larger than the token program, well, ever. But _at least_ with confidence not within 1.16, right?

So as long as we have the ability to increase size in >1.16 should be more than enough, more than likely overkill ha

joncinque commented on runtime/src/bank.rs:
```diff
@@ -8119,6 +8128,36 @@ impl Bank {
         }
     }
 
+    fn replace_account<O, N>(
```
micro-nit: it's best to avoid single characters like `O` for readability, since they can get confused for `0`. maybe use `P` for `previous`?

joncinque commented on runtime/src/bank.rs:
```diff
@@ -8131,26 +8170,79 @@ impl Bank {
+                    // lamports for the empty account that will now house the
+                    // PDA
+                    if new_account.lamports() >= lamports {
+                        if let Some(change_in_cap) = new_account.lamports().checked_sub(lamports) {
```
You can just use a `saturating_sub` here since you already checked that `new_account.lamports() >= lamports`

joncinque commented on runtime/src/bank.rs:
```diff
@@ -8131,26 +8170,79 @@ impl Bank {
+                            self.replace_account(
+                                old_address,
+                                new_address,
+                                self.get_account_with_fixed_root(old_address).as_ref(),
```
You can just pass `None` here since you already fetched it earlier

joncinque reviewed (commented):
This is pretty much ready!

Could you also test this in a local `solana-test-validator` to make sure that the feature works as expected? Ie. you build your own solana-test-validator with the new feature, run it with the feature forced off and short epochs, upload the new program to the new address, activate the feature, then see that the no-op program behaves as expected on address `Feature11111...` when you send instructions to it

buffalojoec commented:
> Could you also test this in a local `solana-test-validator` to make sure that the feature works as expected? Ie. you build your own solana-test-validator with the new feature, run it with the feature forced off and short epochs, upload the new program to the new address, activate the feature, then see that the no-op program behaves as expected on address `Feature11111...` when you send instructions to it

Do you want me to include these steps to reproduce the test within this PR? Or elsewhere?

buffalojoec committed:
more cleanup

buffalojoec commented on runtime/src/inline_feature_gate_program.rs:
```diff
@@ -0,0 +1,5 @@
+//! Contains replacement program IDs for the feature gate program
+
+pub(crate) mod noop_program {
+    solana_sdk::declare_id!("5Re2FUHmvSdem1KV2bU5GFvqZpP2MXj76yVhMKKbP45o");
```
Just putting a reminder here for myself once the verifiable build stuff is done. Replace with new program ID

joncinque commented:
Summoning @mvines to take a look! Feel free to add anyone else who might want to review. There's just the new program id to tweak, then this should be good to go

buffalojoec committed:
add verifiably built program

buffalojoec commented:
> > Do you want me to include these steps to reproduce the test within this PR? Or elsewhere?
> 
> Nah don't worry about it, it's more of an end-to-end test to make sure this is doing exactly what we expect

I'll work on this now. I've committed the verifiably-built program ID

buffalojoec committed:
update program account state

buffalojoec commented:
The run-through with the test validator revealed some suspect behavior.

> Note I used a different program ID than the verifiably-built one provided in this PR



Consider the two program IDs:

```text
7w1PS4TpMEYFwZcRn1U2xv7T8f9hjvw2hmddazwa1JEm
Feature111111111111111111111111111111111111
```

And their data accounts:

```text
$ solana find-program-derived-address BPFLoaderUpgradeab1e11111111111111111111111 pubkey:7w1PS4TpMEYFwZcRn1U2xv7T8f9hjvw2hmddazwa1JEm

ASA6aJacK1mdQbK94BFMbovhrYoH474uegksKBaqeEoU

$ solana find-program-derived-address BPFLoaderUpgradeab1e11111111111111111111111 pubkey:Feature111111111111111111111111111111111111

H253PPA38ecQvnpSvfWgqBfMuVgZ656CEXLgTHwZQHfs
```

The new program account comes back missing 4 bytes.

```text
$ diff <(solana account -ul 7w1PS4TpMEYFwZcRn1U2xv7T8f9hjvw2hmddazwa1JEm) <(solana account -ul Feature111111111111111111111111111111111111)
2,3c2,3
< Public Key: 7w1PS4TpMEYFwZcRn1U2xv7T8f9hjvw2hmddazwa1JEm
< Balance: 0.00114144 SOL
---
> Public Key: Feature111111111111111111111111111111111111
> Balance: 0.0011136 SOL
7,10c7,9
< Length: 36 (0x24) bytes
< 0000:   02 00 00 00  8c 2a fc d1  0a dd 1a cd  a6 fa 41 aa   .....*........A.
< 0010:   41 30 cb 09  dc 72 53 75  75 f7 d2 f7  74 e3 64 b0   A0...rSuu...t.d.
< 0020:   d3 fa 89 43                                          ...C
---
> Length: 32 (0x20) bytes
> 0000:   ee 01 24 15  d1 7d 3d 15  18 93 05 40  18 8b aa cb   ..$..}=....@....
> 0010:   a7 58 f5 a4  3e fe bd 8f  c0 35 03 8f  79 27 20 66   .X..>....5..y' f
```

The data accounts have slightly different bytes in the first few indices of the buffer.

```text
$ diff <(solana account -ul ASA6aJacK1mdQbK94BFMbovhrYoH474uegksKBaqeEoU) <(solana account -ul H253PPA38ecQvnpSvfWgqBfMuVgZ656CEXLgTHwZQHfs)
2c2
< Public Key: ASA6aJacK1mdQbK94BFMbovhrYoH474uegksKBaqeEoU
---
> Public Key: H253PPA38ecQvnpSvfWgqBfMuVgZ656CEXLgTHwZQHfs
8c8
< 0000:   03 00 00 00  67 07 00 00  00 00 00 00  01 14 bc 3e   ....g..........>
---
> 0000:   03 00 00 00  30 00 00 00  00 00 00 00  01 14 bc 3e   ....0..........>
```

The first issue I can trace to the [state of a program account that gets written by the BPF Upgradeable Loader](https://github.com/solana-labs/solana/blob/cd96d5b6c3fcb80905648261567e893a94e16f5a/programs/bpf_loader/src/lib.rs#L761). As you can see from the link, it's not just the raw bytes of the PDA, but a struct serialized.

I've remedied this problem in commit [92b4268](https://github.com/solana-labs/solana/pull/32783/commits/92b4268fe93123819a10b240bf77250526bfd1c4) and I've also included renaming of `replace_program_account` so engineers know that function is explicitly for non-upgradeable programs.

After commit [92b4268](https://github.com/solana-labs/solana/pull/32783/commits/92b4268fe93123819a10b240bf77250526bfd1c4), I ran through it again and the program account looks normal:

```text
diff <(solana account -ul 7w1PS4TpMEYFwZcRn1U2xv7T8f9hjvw2hmddazwa1JEm) <(solana account -ul Feature111111111111111111111111111111111111)
2c2
< Public Key: 7w1PS4TpMEYFwZcRn1U2xv7T8f9hjvw2hmddazwa1JEm
---
> Public Key: Feature111111111111111111111111111111111111
8,10c8,10
< 0000:   02 00 00 00  8c 2a fc d1  0a dd 1a cd  a6 fa 41 aa   .....*........A.
< 0010:   41 30 cb 09  dc 72 53 75  75 f7 d2 f7  74 e3 64 b0   A0...rSuu...t.d.
< 0020:   d3 fa 89 43                                          ...C
---
> 0000:   02 00 00 00  ee 01 24 15  d1 7d 3d 15  18 93 05 40   ......$..}=....@
> 0010:   18 8b aa cb  a7 58 f5 a4  3e fe bd 8f  c0 35 03 8f   .....X..>....5..
> 0020:   79 27 20 66                                          y' f
```

But the data account still has those different first few bytes:

```text
diff <(solana account -ul ASA6aJacK1mdQbK94BFMbovhrYoH474uegksKBaqeEoU) <(solana account -ul H253PPA38ecQvnpSvfWgqBfMuVgZ656CEXLgTHwZQHfs)
2c2
< Public Key: ASA6aJacK1mdQbK94BFMbovhrYoH474uegksKBaqeEoU
---
> Public Key: H253PPA38ecQvnpSvfWgqBfMuVgZ656CEXLgTHwZQHfs
8c8
< 0000:   03 00 00 00  09 02 00 00  00 00 00 00  01 14 bc 3e   ...............>
---
> 0000:   03 00 00 00  36 00 00 00  00 00 00 00  01 14 bc 3e   ....6..........>
```

It looks like two bytes' difference. I think there might be some bytes at the beginning of the data account that are specific to the program account's ID. Investigating

buffalojoec commented:
The feature account even comes up as a program on explorer!

<img width="1233" alt="Screenshot 2023-08-22 at 4 05 24 PM" src="https://github.com/solana-labs/solana/assets/67205063/1c9cc139-4958-4cea-876d-902eb07717b7">

buffalojoec commented:
Experimenting with this some more, I can see that if I deploy just two random programs, they _also_ differ in these particular bytes.

```text
diff <(solana account -ul D9fxtyyQgxGr7sTzgSt2XiJe2BjJouebPpuVCQiRNeej) <(solana account -ul HQ7mntpJyNC3BjXWiWzw26ttcPZxGE7QTSdJ7Gx2VadF)
2c2
< Public Key: D9fxtyyQgxGr7sTzgSt2XiJe2BjJouebPpuVCQiRNeej
---
> Public Key: HQ7mntpJyNC3BjXWiWzw26ttcPZxGE7QTSdJ7Gx2VadF
8c8
< 0000:   03 00 00 00  1c 00 00 00  00 00 00 00  01 14 bc 3e   ...............>
---
> 0000:   03 00 00 00  55 00 00 00  00 00 00 00  01 14 bc 3e   ....U..........>

```

I've deduced this to be the [current slot of the deployment](https://github.com/solana-labs/solana/blob/cd96d5b6c3fcb80905648261567e893a94e16f5a/programs/bpf_loader/src/lib.rs#L733).

My issue previously was I was re-deploying the program to validate the data, not realizing that it would in fact bear a new deployment slot.

When I deploy the program and inspect the bytes manually (or using files), we can see they are the same.

```text
diff noop_program.txt feature_program.txt                                                                                                 
2c2
< Public Key: D9fxtyyQgxGr7sTzgSt2XiJe2BjJouebPpuVCQiRNeej
---
> Public Key: H253PPA38ecQvnpSvfWgqBfMuVgZ656CEXLgTHwZQHfs
```

buffalojoec commented:
Sorry for all the comments, but now this is ready for review (again).

buffalojoec committed:
cleaned up serializing logic

joncinque reviewed (dismissed):
I double-checked the deployed program, and it matches as expected. Everything looks good to me! Let's get one more approval to be sure

mvines commented on runtime/src/inline_feature_gate_program.rs:
```diff
@@ -0,0 +1,5 @@
+//! Contains replacement program IDs for the feature gate program
+
+pub(crate) mod noop_program {
+    solana_sdk::declare_id!("2rqZsQBbacRbuAuTSuJ7n49UQT9fzes8RLggFcmB9YuN");
```
As a part of this, should we start including the new feature-gate program in:
1. https://github.com/solana-labs/solana/blob/master/fetch-spl.sh
2. https://github.com/solana-labs/solana/blob/master/program-test/src/programs.rs
for new clusters?

buffalojoec replied:
Yeah, probably! Since it's not a native program. 

With that being said, is it weird that we'd have cluster operations dependent on a non-native program?

mvines replied:
In this case I think it's ok. But by adding the program elf in those two places, the program should be added to new genesis configs that are created when a test cluster is booted, so the program will always be present practically

buffalojoec replied:
I think in order to add the program here, we'd have to merge [this PR](https://github.com/solana-labs/solana-program-library/pull/4987) and then publish a release for the Feature Gate program, right?

mvines commented on runtime/src/bank.rs:
```diff
@@ -8119,9 +8128,39 @@ impl Bank {
 
+    fn replace_account<U, V>(
+        &mut self,
+        old_address: &Pubkey,
+        new_address: &Pubkey,
+        old_account: Option<&U>,
+        new_account: &V,
```
I found the usage of "old" and "new" here confusing. This function also does more than a replacement, it also deletes an account.

I feel it would be more readable if named like `move_account(src_address, src_account, dst_address, dst_account)`

mvines commented on runtime/src/bank.rs:
```diff
@@ -8131,26 +8170,87 @@ impl Bank {
+    ) {
+        // Must be attempting to replace an empty account with a program
+        // account _and_ data account
+        if let Some(new_account) = self.get_account_with_fixed_root(new_address) {
```
emit a `warn!` if the account is `None`?

mvines commented on runtime/src/bank.rs:
```diff
@@ -8131,26 +8170,87 @@ impl Bank {
+                &[new_address.as_ref()],
+                &bpf_loader_upgradeable::id(),
+            );
+            if let Some(new_data_account) = self.get_account_with_fixed_root(&new_data_address) {
```
emit a `warn!` if the account is `None`?

mvines commented on runtime/src/bank.rs:
```diff
@@ -8131,26 +8170,87 @@ impl Bank {
+    /// Use to replace an empty account with a program by feature activation
+    fn replace_empty_account_with_upgradeable_program(
+        &mut self,
+        old_address: &Pubkey,
+        new_address: &Pubkey,
```
old/new trips me up here too

mvines commented on runtime/src/bank.rs:
```diff
@@ -8131,26 +8170,87 @@ impl Bank {
+                    let state = UpgradeableLoaderState::Program {
+                        programdata_address: old_data_address,
+                    };
+                    if let Ok(data) = bincode::serialize(&state) {
```
logging an error if this ever fails would be nice too

buffalojoec committed:
use full word capitalization

buffalojoec committed:
rename old & new to dst & src

buffalojoec committed:
swap src and dst in parameters

buffalojoec committed:
add warnings and errors

buffalojoec commented:
@mvines addressed all feedback except the elf files comment, in separate commits

t-nelson commented on sdk/src/feature_set.rs:
```diff
@@ -687,6 +687,10 @@ pub mod reduce_stake_warmup_cooldown {
     }
 }
 
+pub mod feature_gate_program {
```
"verb" + noun is a bit more descriptive
```suggestion
pub mod programify_feature_gate_program {
```

t-nelson commented on sdk/src/feature_set.rs:
```diff
@@ -851,6 +855,7 @@ lazy_static! {
         (bpf_account_data_direct_mapping::id(), "use memory regions to map account data into the rbpf vm instead of copying the data"),
         (last_restart_slot_sysvar::id(), "enable new sysvar last_restart_slot"),
         (reduce_stake_warmup_cooldown::id(), "reduce stake warmup cooldown from 25% to 9%"),
+        (feature_gate_program::id(), "control feature activations with an on-chain program #32783"),
```
to improve accuracy of the change

```suggestion
        (programify_feature_gate_program::id(), "move feature gate activation logic to an on-chain program #32783"),
```

t-nelson reviewed (commented):
this looks to be on the right track. my concerns are mostly with propagating pre-existing bad habits 😉

buffalojoec committed:
rename feature to programify

buffalojoec committed:
test suite description clarity

buffalojoec commented on runtime/src/bank.rs:
```diff
@@ -8119,39 +8128,227 @@ impl Bank {
-                    .remove_programs([*old_address].into_iter());
+                    .remove_programs([*dst_address].into_iter());
+            } else {
+                warn!("Unable to find source program {}", src_address);
```
What's best practice for handling fallible functions in feature activations?

It seemed like previously people were panicking and unwrapping, and I don't see a whole lot of `Result` handling, especially in the infallible `apply_feature_activations`.

Are you suggesting make these "replace" functions return `Result`s and handle them within the feature activation gracefully? Assuming just kind of log it all but continue with the epoch rollover not panicking?

Fwiw, I sort of followed the last one for the nonupgradeable which was just one `unwrap()`

buffalojoec committed:
remove strings from datapoints

buffalojoec commented on runtime/src/bank.rs:
```diff
@@ -8119,39 +8128,227 @@ impl Bank {
     ) {
-        if let Some(old_account) = self.get_account_with_fixed_root(old_address) {
-            if let Some(new_account) = self.get_account_with_fixed_root(new_address) {
+        if let Some(dst_account) = self.get_account_with_fixed_root(dst_address) {
+            if let Some(src_account) = self.get_account_with_fixed_root(src_address) {
```
Yeah, I agree. I'll add this in once I understand how you want to handle the errors better

buffalojoec commented:
@t-nelson addressed some of the feedback, and left a few questions for you. Thanks!

CriesofCarrots reviewed (changes requested):
~~The issue I mentioned on bank.rsL8221 is a blocker.~~ But the approach looks generally good!

As someone coming in to review late in the discussion, I would have had an easier time with a smaller set of more-focused commits. Something like:
1. Split `move_account` logic out of `replace_program_account()` (to verify no change in behavior)
2. Update account names in `replace_program_account()`
3. Add `replace_empty_account_with_upgradeable_program()`
4. Add tests
5. Add feature

Just something to consider for the future; maybe rewrite commit history when a feature goes from draft/dev to review status?

CriesofCarrots commented on runtime/src/bank.rs:
```diff
@@ -8119,39 +8128,168 @@ impl Bank {
+                // We aren't going to check that there isn't a data account at
+                // the known program-derived address (ie. `dst_data_address`),
+                // because if it exists, it will be overwritten
+                if self.get_account_with_fixed_root(dst_address).is_none() {
```
Hey hey! Sorry about the whiplash, but I was actually wrong about this. The feature program is the one native program that *isn't* populated on chain. So the account is indeed empty and your tests are fine.

buffalojoec committed:
spell out source and destination

buffalojoec committed:
more verbose comments in account replace functions

buffalojoec committed:
move lamport calculation

buffalojoec committed:
swap lamport check for state check

buffalojoec committed:
move replace functions to helper module

buffalojoec commented on runtime/src/bank.rs:
```diff
@@ -8119,39 +8128,168 @@ impl Bank {
+                        // lamports for the empty account that will now house the
+                        // PDA, and we can properly serialize the program account's
+                        // state
+                        if src_account.lamports() >= lamports {
```
I guess you're right, we shouldn't ever. But this was just to be _super_ safe. 

I'll add a check for the data type instead!

buffalojoec commented:
Thanks for the review @CriesofCarrots !

> Just something to consider for the future; maybe rewrite commit history when a feature goes from draft/dev to review status?

I can and will definitely do this in the future! I was trying to preserve the timeline with comments and changes, etc. but that makes total sense to me!

I've addressed each comment with its own commit. I can also rebase when we're all set.

I was wondering if you might have some insight into how to make these functions fallible and properly handle them in the feature activation function, as per @t-nelson's comment here: https://github.com/solana-labs/solana/pull/32783#discussion_r1327767028

CriesofCarrots commented:
> I've addressed each comment with its own commit. I can also rebase when we're all set.

Awesome. The changes look good, although clippy has one nit, and you're in feature_set rebase hell 😕 

> I was wondering if you might have some insight into how to make these functions fallible and properly handle them in the feature activation function, as per @t-nelson's comment here: [#32783 (comment)](https://github.com/solana-labs/solana/pull/32783#discussion_r1327767028)

The caller `Bank::apply_feature_activations()` is *not* fallible, and it seems out of the scope of this PR to handle all the plumbing to make it so. Therefore, what I would recommend is to make the methods in `bank::replace_account` fallible -- ie. return `Result<(), ReplaceAccountError>`, where ReplaceAccountError is a new error enum in that module -- and handle the `warn!` and `datapoint_warn!` in a single place in bank.rsL7979. Anything upstream of that would need to be part of a larger Bank refactor, imo. How does that sound?

buffalojoec commented:
> How does that sound?

That sounds great! I'll do that. And yep, I'll rebase too 🪦

buffalojoec committed:
make replace_account methods fallible

buffalojoec commented:
Done!

CriesofCarrots commented on runtime/src/bank/replace_account.rs:
```diff
@@ -0,0 +1,215 @@
+                                    .fetch_sub(change_in_capitalization, Relaxed);
+
+                                Ok(())
+                            } else {
+                                Err(ReplaceAccountError::UnableToSerializeProgramAccountState)
+                            }
+                        } else {
+                            Err(ReplaceAccountError::DestinationAccountExists)
+                        }
+                    } else {
+                        Err(ReplaceAccountError::SourceDataAccountNotFound)
+                    }
+                }
+                _ => Err(ReplaceAccountError::NotAnUpgradeableProgram),
+            }
+        } else {
+            Err(ReplaceAccountError::NotAnUpgradeableProgram)
+        }
+    } else {
+        Err(ReplaceAccountError::SourceAccountNotFound)
+    }
+}
```
Same comment about removing `if` cases 😅

CriesofCarrots reviewed (commented):
Great! warning handling in bank.rs looks good. Some suggestions to make the replace_account methods more readable, plus the clippy lint. Then I think this is good to land!

mvines commented:
We'd definitely want to get a v1.17 backport for this guy, so the sooner we can get this landed the better 🥺

buffalojoec committed:
refactor error handling

buffalojoec commented:
> Great! warning handling in bank.rs looks good. Some suggestions to make the replace_account methods more readable, plus the clippy lint. Then I think this is good to land!

Nice suggestions! Made things _much_ cleaner.

buffalojoec committed:
add test for source program state

CriesofCarrots reviewed (approved):
Thanks for all the polish!

jstarry commented on runtime/src/bank.rs:
```diff
@@ -8054,6 +8056,24 @@ impl Bank {
+            let datapoint_name = "bank-progamify_feature_gate_program";
+            if let Err(e) = replace_account::replace_empty_account_with_upgradeable_program(
+                self,
+                &feature::id(),
+                &inline_feature_gate_program::noop_program::id(),
```
Are these arguments swapped? My impression is that this code is trying to copy code from somewhere into the account with id of `feature::id()` but this code uses `feature_id()` as the `source` instead of the `destination`.

buffalojoec replied:
They do appear to be swapped. Great catch, thanks! 😅

#33894

jstarry commented:
I personally don't think this approach is a good idea. I think the feature program should be added as a builtin program rather than an upgradeable program. The feature program is very sensitive and managing the upgrade authority for it sounds like a headache (speaking of, I'm guessing the upgrade authority keypair (4ktVfu5Kfw6kWYRhVjLfBhpEicgX8XekXLZgmooJm5Tu) is currently owned by @buffalojoec?). I think it's better to update the program using feature gates. That way we don't need to verify the bytecode every time it's upgraded.. and if the program is a builtin program, everyone can read the code easily.

This approach also adds a lot of scary methods like `Bank::move_account`, `Bank::replace_empty_account_with_upgradeable_program`, and `Bank::replace_non_upgradeable_program_account` which don't have adequate test coverage right now and could easily be misused in the future. These code paths are not on the hot path so it's easy for other contributors to forget they exist when adding/changing behavior on the hot path.

I strongly think this change should be reverted.

buffalojoec commented:
> I personally don't think this approach is a good idea. I think the feature program should be added as a builtin program rather than an upgradeable program.
...

Some fair points for sure, thanks for weighing in!

I've actually pulled the feature activation from the schedule in order to open up the conversation on this exact concern to a SIMD. We had similar feedback last week on the community call and in some of the [SIMD 0072](https://github.com/solana-foundation/solana-improvement-documents/pull/72) reviews for the architecture this PR sets up a base for.

I'd love to have some more discussion on your thoughts here in [SIMD 0077](https://github.com/solana-foundation/solana-improvement-documents/pull/77), which corresponds to this change!

jstarry commented:
I can weigh in later, but for now, I'd like these changes to be reverted due to the lack of test coverage for the new bank methods added here. Any objection to that?

buffalojoec commented:
> I can weigh in later, but for now, I'd like these changes to be reverted due to the lack of test coverage for the new bank methods added here. Any objection to that?

No, not from me! Seems like a good idea considering we have to wait for the SIMD to be approved now anyway.

jstarry commented:
Yeah that's fair too, thank you!
